### ISSUE TITLE
Implement Vertamedia macros

### ISSUE THREAD

**Issue by lmangani** [lmangani/cLoki#20]
The popular vertamedia plugin for Grafana features macro processors for query.
We should adopt those in cLoki for a unified clickhouse experience.

https://github.com/Vertamedia/clickhouse-grafana/blob/master/src/sql_query.ts

**Comment by Slach** [comment]
as mantainer of `clickhouse-grafana`
I not sure it's a good idea

these macros designed for adapt and bring traditional clickhouse SQL queries to grafana "time-series specific" behavior (like $columns, $rate, etc)

Loki contains own query language and users will expect from cLoki "drop-in" replacement
i think implements all aspect of Loki query language, should more priority

**Comment by lmangani** [comment]
Thanks @Slach - that's exactly the point. We'd like to support pseudo-loki queries, as well as our own macro commands to return data using the Loki connectors. This is already the case right now, ie: 
```
sum by (ruri_user) (rate(duration[300])) from my_database.my_table where duration > 10
```
This is inspired by the `$colums` macro and inverted to produce tag grouping rather than time grouping. Seems to work well and I feel the vertamedia plugin users could optionally bring that syntax in Loki to return tables or timeseries. 

Would love to get your full opinion on this, perhaps we could arrange a meeting sometime to discuss?

**Comment by lmangani** [comment]
@Slach would you be willing to give us some generic feedback on the current cloki macro query approach? 
https://github.com/lmangani/cLoki/wiki/ClickHouse-Metrics

**Comment by Slach** [comment]
@lmangani I still not sure about add additional functionality query language dialect is a good idea, 
sorry guys, maybe I not the right person for questions

could you clarify one thing for me
how you going to insert log data to ClickHouse side?
is your plugin only read-only for exists ClickHouse data?

**Comment by lmangani** [comment]
@Slach thanks! 

cLoki can currently do a few things:

- read from any clickhouse table (no matter how you insert) using the custom [cloki functions](https://github.com/lmangani/cLoki/wiki/ClickHouse-Metrics) and slowly emulate Loki 2.0 style queries on top of that.
- insert logs / fingerprints in clickhouse using the emulated [Loki HTTP API](https://github.com/lmangani/cLoki#api-examples) using [this schema](https://github.com/lmangani/cLoki/wiki/ClickHouse-Schema)
- insert metrics using InfluxDB line protocol (from Telegraf, etc) documented [here](https://github.com/lmangani/cLoki/wiki/Telegraf-HTTP-Input) which is a work in progress.

**Comment by Slach** [comment]
@lmangani ok. so maybe `clickhouse()` still not a good idea, cause it will hard switch between native Loki and cLoki data sources

**Comment by lmangani** [comment]
@Slach good point - will keep it in mind and probably let `clickhouse()` just handle raw clickhouse queries instead.
For clarity, cLoki does not speak or connect to native Loki - it just emulates its API and provide additional functions in-the-middle.
